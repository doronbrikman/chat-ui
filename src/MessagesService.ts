const EVENT = 'spotim/chat';

export type Message = {
  avatar: string;
  username: string;
  text: string;
  timestamp?: number;
};

export class MessagesService {
  constructor(private socket: SocketIOClient.Socket) {}

  connect() {
    this.socket.open();
  }

  sendMessage(msg: Message) {
    msg.timestamp = Date.now();
    this.socket.emit(EVENT, msg);
  }

  subscribeToMessages(cb: (msg: Message) => void) {
    this.socket.on(EVENT, cb);
  }

  disconnect() {
    this.socket.close();
  }
}
