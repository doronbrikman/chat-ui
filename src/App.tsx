import React from 'react';

import { MessagesService, Message as MessageType } from './MessagesService';
import { UserStore, User } from './user';
import { MessageStream } from './MessagesStream';
import { MessageInput } from './MessageInput';
import { MessagesList } from './MessagesList';
import { avatars } from './utils';

export type State = {
  user: User;
};

type Props = {
  messagesService: MessagesService;
};

export class App extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.props.messagesService.connect();

    const user = new UserStore();
    if (user.avatar === '') {
      const randomIndex = Math.floor(Math.random() * 5);
      user.avatar = avatars[randomIndex];
    }

    this.state = { user };
  }

  handleUsernameChange = (username: string) => {
    const user = this.state.user;
    user.username = username;

    this.setState({ user });
  };

  componentWillUnmount() {
    this.props.messagesService.disconnect();
  }

  render() {
    const { messagesService } = this.props;
    const { user } = this.state;

    return (
      <div
        style={{
          display: 'flex',
          flexDirection: 'column',
          height: '100vh',
          maxWidth: 500,
          margin: '0 auto',
          padding: 4
        }}
      >
        <div style={{ display: 'flex', flex: 1, padding: 4 }}>
          <MessageStream service={messagesService}>
            {messages => <MessagesList user={user} messages={messages} />}
          </MessageStream>
        </div>
        <div>
          <MessageInput
            user={user}
            onUsernameChange={this.handleUsernameChange}
            onSubmit={(msg: MessageType) => messagesService.sendMessage(msg)}
          />
        </div>
      </div>
    );
  }
}
