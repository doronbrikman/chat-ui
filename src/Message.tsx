import React from 'react';
import { Message, Image } from 'semantic-ui-react';

import { Message as MessageType } from './MessagesService';

type Props = {
  isMe: boolean;
  message: MessageType;
};

export const ChatMessage: React.StatelessComponent<Props> = ({
  isMe,
  message: { username, avatar, text }
}) => (
  <Message icon color={isMe ? 'green' : 'blue'}>
    <Image src={avatar} size="mini" />
    <Message.Content>
      <Message.Header>{username}</Message.Header>
      {text}
    </Message.Content>
  </Message>
);
