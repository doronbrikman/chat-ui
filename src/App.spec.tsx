import React from 'react';
import { mount } from 'enzyme';
import io from 'socket.io-client';

import { App } from './App';
import { MessagesService, Message } from './MessagesService';
import { ChatMessage } from './Message';

jest.mock('socket.io-client');

jest.mock('./user');

test('renders correctly', () => {
  const socket = io('');
  const messagesService = new MessagesService(socket);

  const wrapper = mount(<App messagesService={messagesService} />);
  expect(wrapper).toMatchSnapshot();
});

test('renders 1 message', () => {
  const socket = io('');
  const messagesService = new MessagesService(socket);

  const message: Message = {
    username: 'doron',
    avatar: 'img',
    text: 'hello'
  };

  // mocking sending a message to server
  messagesService.sendMessage(message);

  const wrapper = mount(<App messagesService={messagesService} />);
  expect(wrapper.find(ChatMessage)).toBeTruthy();
});
