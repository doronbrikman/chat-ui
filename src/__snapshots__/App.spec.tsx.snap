// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<App
  messagesService={
    MessagesService {
      "socket": Object {
        "close": [Function],
        "emit": [Function],
        "on": [Function],
        "open": [Function],
      },
    }
  }
>
  <div
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "height": "100vh",
        "margin": "0 auto",
        "maxWidth": 500,
        "padding": 4,
      }
    }
  >
    <div
      style={
        Object {
          "display": "flex",
          "flex": 1,
          "padding": 4,
        }
      }
    >
      <MessageStream
        service={
          MessagesService {
            "socket": Object {
              "close": [Function],
              "emit": [Function],
              "on": [Function],
              "open": [Function],
            },
          }
        }
      >
        <MessagesList
          messages={Array []}
          user={
            UserStore {
              "avatar": "d",
              "username": "d",
            }
          }
        >
          <ul
            onScroll={[Function]}
            style={
              Object {
                "height": "100%",
                "margin": 0,
                "overflow": "auto",
                "padding": 0,
                "width": "100%",
              }
            }
          />
        </MessagesList>
      </MessageStream>
    </div>
    <div>
      <MessageInput
        onSubmit={[Function]}
        onUsernameChange={[Function]}
        user={
          UserStore {
            "avatar": "d",
            "username": "d",
          }
        }
      >
        <div
          style={
            Object {
              "display": "flex",
              "height": 30,
              "justifyContent": "space-around",
            }
          }
        >
          <Image
            as="img"
            size="mini"
            src="d"
            ui={true}
          >
            <img
              className="ui mini image"
              src="d"
            />
          </Image>
          <Input
            onChange={[Function]}
            placeholder="username"
            type="text"
            value="d"
          >
            <div
              className="ui input"
            >
              <input
                onChange={[Function]}
                placeholder="username"
                type="text"
                value="d"
              />
            </div>
          </Input>
          <Input
            onChange={[Function]}
            placeholder="Type Somthing..."
            type="text"
            value=""
          >
            <div
              className="ui input"
            >
              <input
                onChange={[Function]}
                placeholder="Type Somthing..."
                type="text"
                value=""
              />
            </div>
          </Input>
          <Button
            as="button"
            disabled={true}
            onClick={[Function]}
            role="button"
          >
            <button
              className="ui disabled button"
              disabled={true}
              onClick={[Function]}
              role="button"
              tabIndex={-1}
            >
              Send
            </button>
          </Button>
        </div>
      </MessageInput>
    </div>
  </div>
</App>
`;
